src/games/TicTacToe.ts

```diff
@@ -0,0 +1,400 @@
+
+        const ml = game.board[4] || "4";
+        const mm = game.board[5] || "5";
+        const mr = game.board[6] || "6";
+
+        const tl = game.board[7] || "7";
+        const tm = game.board[8] || "8";
+        const tr = game.board[9] || "9";
+
+        game.channel.send(`\`\`\`${tl} | ${tm} | ${tr}\n---------\n${ml} | ${mm} | ${mr}\n---------\n${bl} | ${bm} | ${br}\`\`\``);
+    }
+
+    private handleChat(message: Discord.Message) {
+        const game = this.games.find(g => g.channel.id === message.channel.id);
+        if (!game) return;
+
+        if (game.isActive) {
+            if (message.cleanContent === "!draw") {
+                this.handleDraw(game, message.author.id);
+            }
+
+            if (message.cleanContent === "!ff") {
+                this.handleFF(game, message.author.id);
+            }
+        }
+
+        if (message.cleanContent === "!rematch") {
+            this.handleRematch(game, message.author.id);
+        }
+
+        // ignore messages from the other player
+        if (game.turnPlayer === TicTacToeType.X) {
+            if (message.author.id === game.Oplayer.id) {
+                return;
+            }
+        }
+        if (game.turnPlayer === TicTacToeType.O) {
+            if (message.author.id === game.Xplayer.id) {
+                return;
+            }
+        }
+
+        const regex = /^\d$/;
+        if (regex.test(message.cleanContent)) {
+            const move = +message.cleanContent;
+            if (this.isValidMove(game.board, move)) {
+                this.incrementGameState(game, move);
+            }
+        }
+    }
+
+    private handleRematch(game: TicTacToeGame, sender: string) {
+        if (sender === game.Xplayer.id) {
+            if (game.wantRematch.includes(game.Xplayer.id)) {
+                return;
+            }
+
+            game.channel.send("Game will rematch when both players have entered the command");
+            game.wantRematch.push(game.Xplayer.id);
+        }
+
+        if (sender === game.Oplayer.id) {
+            if (game.wantRematch.includes(game.Oplayer.id)) {
+                return;
+            }
+
+            game.channel.send("Game will rematch when both players have entered the command");
+            game.wantRematch.push(game.Oplayer.id);
+        }
+
+        if (game.wantRematch.length === 2) {
+            clearTimeout(game.deleteTimeout);
+            this.games = this.games.filter(g => g !== game);
+            this.initFreshGame(game.channel, game.Xplayer, game.Oplayer);
+        }
+    }
+
+    private handleFF(game: TicTacToeGame, sender: string) {
+        if (sender === game.Xplayer.id) {
+            this.handleEndOfGame(game, game.Oplayer.id);
+        }
+
+        if (sender === game.Oplayer.id) {
+            this.handleEndOfGame(game, game.Xplayer.id);
+        }
+    }
+
+    private handleDraw(game: TicTacToeGame, sender: string) {
+        if (sender === game.Xplayer.id) {
```

Comment:
Ever considered doing a for loop with 2 entries instead of duplicating code.

This goes for more than just this block